### PR TITLE
MangleProcs following the Itanium spec so they are demangled in the debugger call stack

### PR DESCRIPTION
**Pull request by jmgomez**
image

### CODE REVIEW

**Comment by elcritch** [comment]
@jmgomez any chance this could be utilized with variables as well?

**Comment by jmgomez** [comment]
> @jmgomez any chance this could be utilized with variables as well?

Unfortunately the debuggers dont pick the Itanium mangling scheme for vars. But locals are fine, mangling on parameters is minimal and in globals is not that bad. In the mid term the plan is to build a custom debugger adapter (so we can also opt-in temps)

File: compiler/ccgutils.nim [hunk 1]
```diff
@@ -153,3 +153,63 @@ proc ccgIntroducedPtr*(conf: ConfigRef; s: PSym, retType: PType): bool =
...
+  case t.kind:
+  of tyObject, tyEnum, tyDistinct, tyUserTypeClass, tyGenericParam: 
+    encodeSym(m, t.sym)
+  of tyGenericInst, tyUserTypeClassInst, tyGenericBody:
+    var name = encodeName(t[0].sym.name.s)
+    name.add "I"
+    for i in 1..<t.len - 1: 
+      name.add encodeType(m, t[i])
+    name.add "E"
+    name
```
Comment (Araq) [inline compiler/ccgutils.nim]:
```suggestion
    result = encodeName(t[0].sym.name.s)
    result.add "I"
    for i in 1..<t.len - 1: 
      result.add encodeType(m, t[i])
    result.add "E"
```

File: compiler/ccgutils.nim [hunk 2]
```diff
@@ -153,3 +153,63 @@ proc ccgIntroducedPtr*(conf: ConfigRef; s: PSym, retType: PType): bool =
...
+  #Module::Type
+  var name = s.name.s 
+  if makeUnique:
+    name = makeUnique(m, s, name)
+  "N" & encodeName(s.owner.name.s) & encodeName(name) & "E"
+
+proc encodeType*(m: BModule; t: PType): string = 
+  var kindName = ($t.kind)[2..^1]
+  kindName[0] = toLower($kindName[0])[0]
+  case t.kind:
```
Comment (Araq) [inline compiler/ccgutils.nim]:
```suggestion
  case t.kind
```

File: compiler/ccgutils.nim [hunk 3]
```diff
@@ -153,3 +153,63 @@ proc ccgIntroducedPtr*(conf: ConfigRef; s: PSym, retType: PType): bool =
...
+      case t.kind:
+      of tySequence: encodeName("seq") 
+      else: encodeName(kindName)
+    name.add "I"
+    for i in 0..<t.len:
+      let s = t[i]  
+      if s.isNil: continue
+      name.add encodeType(m, s)
+    name.add "E"
+    name
```
Comment (Araq) [inline compiler/ccgutils.nim]:
```suggestion
    result = 
      case t.kind:
      of tySequence: encodeName("seq") 
      else: encodeName(kindName)
    result.add "I"
    for i in 0..<t.len:
      let s = t[i]  
      if s.isNil: continue
      result.add encodeType(m, s)
    result.add "E"
```

**Pull request merged**